custom-commands/inotify-proxy/README.md

```diff
@@ -0,0 +1,23 @@
+# inotify-proxy to enable file watchers on NFS shares
+
```

Comment:
A little introduction will probably help here, since people don't automatically know what this is about.
```suggestion

On a regular Linux or macOS filesystem, a process can subscribe to notifications about file changes ("inotify" or "fsnotify").  Most IDEs like PhpStorm depend on this notification to update files or rendering when they change. However, the NFS protocol predates the inotify feature, and so does not support it. However, most developers on macOS and Windows use the [nfs_mount_enabled feature](https://ddev.readthedocs.io/en/stable/users/performance/#using-nfs-to-mount-the-project-into-the-web-container) to mount files from the host into the container instead of using the docker-native technique. Performance is much improved, but of course NFS does not support notifications.

(Note that this is not necessary on Linux hosts; NFS is not needed there because the Docker filesystem mount is so fast.)
```

Reply:
BTW, does this work on WIndows? What's the easiest way to test?

Reply:
It's designed to work inside a docker container. IMHO it's only running on linux.

Reply:
I understand, but does this work on a windows host?

Reply:
Never tested that, because we do not work with Windows. 
Theoretical It should also work on a Windows host with NFS share, but I tested it only with MacOS and Linux.

Reply:
I'll test it if I know how to test it. I think one of those other tools has a host-side listener that worked. Still not understanding how this works without a host-side listener.

Reply:
It has a internal dict with the stats of each watched file.
In every watch cycle it checks if the file is changed. If it's changed, the file will be touched which triggers a Inotify event.
It's only necessary to start the tool by this command after "ddev start".

Reply:
So it works by changing the timestamp of the file? That's a pretty slick, for generated files. Seems like a simpler technique than some others.

Reply:
That's the whole magic.
https://github.com/cmuench/inotify-proxy/blob/1fba6da45ab75a77867b22b3f88d5170ad04ff78/internal/watcher/watcher.go#L98

Reply:
I guess that's what I don't understand about the whole process. If a file changes on the container filesystem, its timestamp already changes. And this technique *still* uses NFS to provide that change to the host filesystem.

Reply:
right :-)

Reply:
@rfay writes to NFS originating on the machine that has mounted an NFS export (in this case, the container) still trigger inotify events as expected. Events originating on the host that's providing the NFS export (the host mac system) don't.

@cmuench have you been using this for long periods of time with large projects? This is effectively just polling NFS for updated metadata and we've run into problems with that approach when using e.g. the polling mode in webpack. I'm curious about the overall reliability here. If you change 1000 files on the host, do you get 1000 inotify events in the guest? Same for 10k?

Reply:
I guess people use inotify (and miss it) two different ways. One is changes in the container not resulting in inotify on the host (which I don't think this does anything about) and the other is changes on the host not resulting in inotify to the container (which this does, right?)

Reply:
@cweagans At netz98 the frontend developers use this tools for different file watch tasks in Magento 2 shops.
Is there any php system with more files then Magento 2? 😉
We do not have the use case that there are several thousand files changing at once. The tool has a garbage collector to prevent a memory leak.
My tool is the golang version if this NPM package https://www.npmjs.com/package/inotify-proxy
The NPM package compiles a c program.
If we change 1.000 files, there should be 1.000 change events in the guest. Currently we do not have problems. The tool is also used by other agencies in the Magento eco-system.

Reply:
@rfay If you work directly inside a container then you do not have problems with missing inotify events. The issue is that someone works in e.g. PhpStorm and changes files. The watch task runs inside the container. If NFS used, no inotify event is triggered. Without the inotify event, the watch task inside the container is not seeing any change. That's annoying.
In the Magento world the people use less/sass/grunt and the Magento Cache Cleaner tool.
Without a good solution, ddev has as low acceptance and people work directly on the host machine.

Reply:
Right, but the opposite problem people have is with GUIs that are expecting notifications when running on the host, and doing a build that runs in the container. So for example, they have gulp or compass running in the container and they change css files.

Reply:
I guess what we need to do here is make sure we're clear which problem is being solved. And pioneer best practices for solving it. Because the more you can do your work where things work best (inside the container, Linux) the better off you are.